### PR TITLE
[Bugfix][KVConnector] Disable cross-layer KV blocks for per-token-head quant

### LINKED ISSUE #48412: [Bug]: OffloadingConnector corrupts outputs with per-token-head quantized KV cache (cross-layer allocation lacks scale packing)
Combining the native offloading connector with a per-token-head quantized KV
cache corrupts generations **immediately** — on the first request, before any
offloaded block is ever restored.

| case (H100, Qwen2.5-0.5B, `VLLM_ATTENTION_BACKEND=TRITON_ATTN`) | generate → `reset_prefix_cache()` → generate |
|---|---|
| bf16 KV + OffloadingConnector | outputs identical ✓ |
| `fp8_per_token_head`, no connector | outputs identical and coherent ✓ |
| `fp8_per_token_head` + OffloadingConnector | **first generation already garbage; second differs from first** ✗ |

Observed on `vllm/vllm-openai:v0.23.0`; the involved code paths are unchanged
on current `main` by inspection.

### Root cause

`OffloadingConnector.prefer_cross_layer_blocks` is hardcoded `True`
(`kv_connector/v1/offloading_connector.py`). For single-group models on
backends with `indexes_kv_by_block_stride`, the model runner therefore
allocates a **cross-layer KV cache** (`kv_connector_model_runner_mixin.py`):

```
INFO [kv_connector_model_runner_mixin.py] Allocating a cross layer KV cache of shape (176452, 2, 24, 2, 16, 64)
```

The last dim is exactly `head_size` (64). Per-token-head modes require the
packed cell layout `2 * (head_size + 4)` with inline fp32 scales per
(head, token) — see `_ensure_scale_caches` in
`v1/attention/backends/triton_attn.py`, which carves scale views at byte
offset `hs` inside each content half. Against the cross-layer allocation
(no scale padding), those views alias K/V data, so writing scales corrupts
neighboring cells from the first forward pass.

### Reproduction

```python
from vllm import LLM, SamplingParams
from vllm.config import KVTransferConfig

prompt = "The following is a technical inventory list. " + " ".join(
    f"Item {i}: component alpha-{i} weighs {i*3} grams and costs {i*7} cents."
    for i in range(1, 61)) + " Question: what does Item 7 weigh and cost? Answer:"
sp = SamplingParams(temperature=0.0, max_tokens=24)

llm = LLM(model="Qwen/Qwen2.5-0.5B-Instruct", kv_cache_dtype="fp8_per_token_head",
          enforce_eager=True, enable_prefix_caching=True, gpu_memory_utilization=0.45,
          max_model_len=2048,
          kv_transfer_config=KVTransferConfig(
              kv_connector="OffloadingConnector", kv_role="kv_both",
              kv_connector_extra_config={"spec_name": "CPUOffloadingSpec",
                                         "cpu_bytes_to_use": 2 * 2**30}))
o1 = llm.generate([prompt], sp)[0].outputs[0]
llm.reset_prefix_cache()
o2 = llm.generate([prompt], sp)[0].outputs[0]
print(o1.text)   # garbage, e.g. ' Item\xa0  \n  \n  \n  ��  ...'
print(o2.text)   # different garbage
```

Dropping `kv_transfer_config` (same quant) or switching to bf16 (same
connector) each produce clean, deterministic outputs.

### PR DESCRIPTION
**Pull request by Achyuthan-S**
## Purpose

Fixes #48412.

Combining `OffloadingConnector` with a per-token-head quantized KV cache
(`fp8_per_token_head` / `int8_per_token_head` / `int4_per_token_head`) corrupts
generations on the **first forward pass**, before any offloaded block is ever
restored.

### Root cause

`OffloadingConnector.prefer_cross_layer_blocks` is hardcoded `True`. For
single-group models on backends with `indexes_kv_by_block_stride` (e.g.
`TRITON_ATTN`), the model runner therefore allocates a **cross-layer** KV cache
(`allocate_uniform_kv_caches` in `kv_connector_model_runner_mixin.py`): one
contiguous buffer shared by all layers.

Per-token-head quant stores an inline fp32 scale in a padded tail of each
`(head, token)` cell. The attention backend carves strided fp32 views over
those pad bytes in `_ensure_scale_caches` (`triton_attn.py`), assuming a
**per-layer, contiguous** buffer. Against the interleaved cross-layer
allocation, the assumed block stride is wrong and every layer's scale views
collide at the same offset, so writing scales aliases neighboring K/V data and
corrupts the cache from the first decode.

### Fix

Return `prefer_cross_layer_blocks = False` when the KV cache dtype uses
per-token-head scales, so these models take the **per-layer** offload path
(which lays out each cell's scale pad correctly). Non-per-token-head configs —
plain `fp8` included — are unchanged and still prefer the cross-layer layout.

This is a soft fallback, not a hard failure: offloading still works for these
dtypes via the per-layer path — only the cross-layer transfer optimization is
skipped, and only for per-token-head dtypes. Teaching the cross-layer allocation
and `_ensure_scale_caches` to pack scales in the shared buffer would restore that
optimization, but is a larger, hotter-path change and can be a follow-up.

### Relationship to #48411 (merged)

#48411 fixed the **per-layer** offload *transfer width* — inline scales were
truncated mid-cell on restore. It explicitly deferred the **cross-layer**
allocation incompatibility to #48412, which this PR addresses. With #48411
already in `main`, this PR completes per-token-head + offloading support:
models fall back to the (now-correct) per-layer path instead of the corrupting
cross-layer allocation. No other open PR addresses #48412.

## Test Plan

```bash
pytest tests/v1/kv_connector/unit/offloading_connector/test_cross_layer_preference.py -v
```

cc @Etelis @orozery

### CODE REVIEW

File: vllm/distributed/kv_transfer/kv_connector/v1/offloading_connector.py [hunk 1]
```diff
@@ -49,7 +52,14 @@
 class OffloadingConnector(KVConnectorBase_V1, SupportsHMA):
     @property
     def prefer_cross_layer_blocks(self) -> bool:
-        return True
+        # Per-token-head KV quant carries inline fp32 scales carved from the raw
+        # KV allocation by the attention backend, which assumes a per-layer,
+        # contiguous buffer. The cross-layer (uniform) allocation shares one
+        # interleaved buffer across all layers, so the scale carving aliases KV
+        # data and corrupts the cache. Fall back to the per-layer offload path.
```
Comment (Etelis) [inline vllm/distributed/kv_transfer/kv_connector/v1/offloading_connector.py]: Trim to <=2 lines - the full mechanism is already in the PR description.
```suggestion
        # Per-token-head quant carves inline-scale views that assume per-layer
        # contiguous KV buffers; the cross-layer layout breaks this and corrupts KV.
```

File: vllm/distributed/kv_transfer/kv_connector/v1/offloading_connector.py [hunk 2]
```diff
@@ -49,7 +52,14 @@
...
+        return not kv_cache_uses_per_token_head_scales(
+            self._vllm_config.cache_config.cache_dtype
+        )
```
Comment (Etelis) [inline vllm/distributed/kv_transfer/kv_connector/v1/offloading_connector.py]: This guards only OffloadingConnector, but `NixlConnector.prefer_cross_layer_blocks` is also `True` (and MooncakeStore's can be via config), so the same corruption stays reachable through them. Consider hoisting the check into `use_uniform_kv_cache()` keyed on `kv_cache_spec.kv_quant_mode.is_per_token_head` - one gate covers every connector.

**Review by Etelis — CHANGES_REQUESTED** [review]
Thank you for the fix and the clear root-cause writeup!

Verified e2e on an H100: on `main` this combination generates nondeterministic garbage from the first pass via the cross-layer allocation; with this patch output is token-exact vs a no-offloading reference and restores work after a prefix-cache reset (bf16 keeps the cross-layer path). One structural request inline.

File: vllm/distributed/kv_transfer/kv_connector/v1/offloading_connector.py (hunk 1 above)
Comment (Achyuthan-S) [reply vllm/distributed/kv_transfer/kv_connector/v1/offloading_connector.py]: Trimmed to 2 lines. It now lives at the gate in use_uniform_kv_cache rather than here — the OffloadingConnector override is reverted.

File: vllm/distributed/kv_transfer/kv_connector/v1/offloading_connector.py (hunk 2 above)
Comment (Achyuthan-S) [reply vllm/distributed/kv_transfer/kv_connector/v1/offloading_connector.py]: Moved the check into use_uniform_kv_cache() keyed on kv_cache_spec.kv_quant_mode.is_per_token_head, so OffloadingConnector, NixlConnector, and MooncakeStore all fall back to the per-layer path for these dtypes. Reverted the OffloadingConnector-specific guard. Coverage added in tests/v1/kv_connector/unit/test_kv_cache_layout.py (per-token-head → rejected; NONE / fp8-per-tensor → allowed).

**Comment by Achyuthan-S** [comment]
Pushed an update addressing both comments: the per-token-head check is now a single gate in use_uniform_kv_cache() (keyed on kv_quant_mode.is_per_token_head) instead of a per-connector override, so it covers every connector that prefers cross-layer blocks. offloading_connector.py is reverted to upstream; net change is +4 lines in the mixin plus a test in test_kv_cache_layout.py. Also rebased onto latest main.

**Comment by Etelis** [comment]
LGTM

@orozery

**Comment by orozery** [comment]
Thanks @Achyuthan-S ! Looks good to me.
I don't think we need the tests. Can we remove them?

**Comment by Achyuthan-S** [comment]
Hey @orozery , removed — the PR is now just the gate in use_uniform_kv_cache. End-to-end behavior was verified on H100 in @Etelis's review above.

**Comment by Achyuthan-S** [comment]
The 6 failures are infra (exit status -1 = agent-lost/cancellation; two logs show "Received cancellation signal", pytorch-compilation was mid-pass). No traceback references this change, and the KV-connector lanes (v1-core-plus-kv, v1-others-cpu, nixlconnector-2gpu) all passed. 

Could a maintainer re-trigger the failed jobs?

**Review by orozery — APPROVED** [review]
Thanks @Achyuthan-S @Etelis !
Let's just update branch until it passes CI.

**Comment by Achyuthan-S** [comment]
The remaining v1-sample-plus-logits failure is unrelated: the 5 failures are test_logprobs_mode with gpu_memory_utilization=0.05 and 2 other LLMs alive on a 16 GB MIG slice — the engine core proc dies at startup under memory pressure (Failed core proc(s): {}), free mem dropping 15.7→6.8 GiB across the stacked LLMs. No traceback references this change, it uses no KV connector, and the 294 other logprobs tests + all KV lanes passed. Should be clear on a re-run of that job.

**Comment by Achyuthan-S** [comment]
Small update .

The 24 failures are not from this change (still +4 lines, 1 file; no failing log references it):
- 20/24 are Buildkite infra — the x86 agent-environment hook failed at setup (Cannot perform an interactive login from a non TTY device), so tests never ran.
- 4/24 are pre-existing main regressions in unrelated areas (test_model_arch_config[Zamba2] dtype assert, test_compile_correctness, granite-4.0 test_hybrid, test_prompt_logprobs_e2e shape error) — none touch use_uniform_kv_cache.
